File: synapxion-train/utils/silabificador.py
class SilabificadorFast:

    @staticmethod
    def is_vowel(c: str) -> bool:
        return c in "aeiouáéíóúüAEIOUÁÉÍÓÚÜ"

    @staticmethod
    def is_strong(c: str) -> bool:
        return c in "aáeéoóAÁEÉOÓ"

    @staticmethod
    def is_weak(c: str) -> bool:
        return c in "iíuúüIÍUÚÜ"

    @staticmethod
    def has_accent(c: str) -> bool:
        return c in "áéíóúÁÉÍÓÚ"

    @staticmethod
    def is_inseparable(c1: str, c2: str) -> bool:
        pair = (c1.lower(), c2.lower())
        return pair in {
            ('b','r'), ('b','l'),
            ('c','r'), ('c','l'),
            ('d','r'),
            ('f','r'), ('f','l'),
            ('g','r'), ('g','l'),
            ('p','r'), ('p','l'),
            ('t','r')
        }

    # 🔥 CLON DIRECTO DEL ALGORITMO C#
    @staticmethod
    def split_into_syllables(word: str):
        syllables = []

        start = 0
        i = 0
        length = len(word)

        while i < length:
            c = word[i]
            c_lower = c.lower()

            if not SilabificadorFast.is_vowel(c_lower):
                i += 1
                continue

            next_i = i + 1

            # =========================
            # DIPTONGO / HIATO
            # =========================
            if next_i < length:
                v1 = c_lower
                v2 = word[next_i].lower()

                if SilabificadorFast.is_vowel(v2):
                    hiato = (
                        (SilabificadorFast.is_strong(v1) and SilabificadorFast.is_strong(v2)) or
                        (SilabificadorFast.is_strong(v1) and SilabificadorFast.is_weak(v2) and SilabificadorFast.has_accent(v2)) or
                        (SilabificadorFast.is_weak(v1) and SilabificadorFast.is_strong(v2) and SilabificadorFast.has_accent(v1))
                    )

                    if not hiato:
                        i += 1  # 🔥 saltar como en C#

            # =========================
            # CONTAR CONSONANTES
            # =========================
            consonants = 0
            j = i + 1

            while j < length and not SilabificadorFast.is_vowel(word[j].lower()):
                consonants += 1
                j += 1

            # =========================
            # REGLAS DE CORTE
            # =========================
            if consonants == 1:
                syllables.append(word[start:i+1])
                start = i + 1

            elif consonants == 2:
                c1 = word[i + 1].lower()
                c2 = word[i + 2].lower()

                if SilabificadorFast.is_inseparable(c1, c2):
                    syllables.append(word[start:i+1])
                    start = i + 1
                else:
                    syllables.append(word[start:i+2])
                    start = i + 2
                    i += 1

            elif consonants > 2:
                syllables.append(word[start:i+2])
                start = i + 2
                i += 1

            i += 1

        # =========================
        # RESTO FINAL
        # =========================
        if start < length:
            syllables.append(word[start:])

        return syllables

File: synapxion-utils/Utils/Silabificador.cs
using System;
using System.Runtime.CompilerServices;

namespace Rudwolf.Utils
{
    public static class SilabificadorFast
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static bool IsVowel(char c)
            => c switch
            {
                'a' or 'e' or 'i' or 'o' or 'u' or
                'á' or 'é' or 'í' or 'ó' or 'ú' or 'ü' => true,
                _ => false
            };

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static bool IsStrong(char c)
            => c is 'a' or 'á' or 'e' or 'é' or 'o' or 'ó';

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static bool IsWeak(char c)
            => c is 'i' or 'í' or 'u' or 'ú' or 'ü';

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static bool HasAccent(char c)
            => c is 'á' or 'é' or 'í' or 'ó' or 'ú';

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static bool IsInseparable(char c1, char c2)
            => (c1, c2) switch
            {
                ('b', 'r') or ('b', 'l') or
                ('c', 'r') or ('c', 'l') or
                ('d', 'r') or
                ('f', 'r') or ('f', 'l') or
                ('g', 'r') or ('g', 'l') or
                ('p', 'r') or ('p', 'l') or
                ('t', 'r') => true,
                _ => false
            };

        // 🔥 CORE: sin listas, sin strings intermedios
        public static void SplitIntoSyllables(
            ReadOnlySpan<char> word,
            Action<ReadOnlySpan<char>> onSyllable)
        {
            int start = 0;

            for (int i = 0; i < word.Length; i++)
            {
                char c = word[i];
                char cLower = char.ToLowerInvariant(c);

                if (!IsVowel(cLower))
                    continue;

                int next = i + 1;

                // Diptongo / hiato
                if (next < word.Length)
                {
                    char v1 = cLower;
                    char v2 = char.ToLowerInvariant(word[next]);

                    if (IsVowel(v2))
                    {
                        bool hiato =
                            (IsStrong(v1) && IsStrong(v2)) ||
                            (IsStrong(v1) && IsWeak(v2) && HasAccent(v2)) ||
                            (IsWeak(v1) && IsStrong(v2) && HasAccent(v1));

                        if (!hiato)
                            i++; // saltar
                    }
                }

                int consonants = 0;
                int j = i + 1;

                while (j < word.Length && !IsVowel(char.ToLowerInvariant(word[j])))
                {
                    consonants++;
                    j++;
                }

                if (consonants == 1)
                {
                    Emit(word, start, i, onSyllable);
                    start = i + 1;
                }
                else if (consonants == 2)
                {
                    char c1 = char.ToLowerInvariant(word[i + 1]);
                    char c2 = char.ToLowerInvariant(word[i + 2]);

                    if (IsInseparable(c1, c2))
                    {
                        Emit(word, start, i, onSyllable);
                        start = i + 1;
                    }
                    else
                    {
                        Emit(word, start, i + 1, onSyllable);
                        start = i + 2;
                        i++;
                    }
                }
                else if (consonants > 2)
                {
                    Emit(word, start, i + 1, onSyllable);
                    start = i + 2;
                    i++;
                }
            }

            if (start < word.Length)
                onSyllable(word[start..]);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static void Emit(
            ReadOnlySpan<char> word,
            int start,
            int end,
            Action<ReadOnlySpan<char>> output)
        {
            output(word.Slice(start, end - start + 1));
        }
    }
}